ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/IndexStreamingCurrentLedgerConfiguration.scala

```diff
@@ -26,46 +27,36 @@ import scala.concurrent.{ExecutionContext, Future, Promise}
   * multiple services and validators require the latest ledger config.
   */
 private[apiserver] final class IndexStreamingCurrentLedgerConfiguration private (
-    index: IndexConfigManagementService,
     ledgerConfiguration: LedgerConfiguration,
+    index: IndexConfigManagementService,
+    scheduler: Scheduler,
+)(implicit
     materializer: Materializer,
-)(implicit loggingContext: LoggingContext)
-    extends CurrentLedgerConfiguration
+    executionContext: ExecutionContext,
+    loggingContext: LoggingContext,
+) extends CurrentLedgerConfiguration
     with AutoCloseable {
 
-  private[this] val logger = ContextualizedLogger.get(this.getClass)
+  private val logger = ContextualizedLogger.get(this.getClass)
 
   // The latest offset that was read (if any), and the latest ledger configuration found (if any)
-  private[this] type StateType = (Option[LedgerOffset.Absolute], Option[Configuration])
-  private[this] val latestConfigurationState: AtomicReference[StateType] =
-    new AtomicReference(None -> None)
-  private[this] val killSwitch: AtomicReference[Option[UniqueKillSwitch]] =
-    new AtomicReference(None)
-  private[this] val readyPromise: Promise[Unit] = Promise()
+  private type StateType = (Option[LedgerOffset.Absolute], Option[Configuration])
+  private val latestConfigurationState = new AtomicReference[StateType](None -> None)
+  private val killSwitch = new AtomicReference[Option[UniqueKillSwitch]](None)
+  private val readyPromise = Promise[Unit]()
 
   // At startup, do the following:
   // - Start loading the ledger configuration
-  // - Mark the provider as ready if no configuration was found after a timeout
   // - Submit the initial config if none is found after a delay
+  // - Mark the provider as ready if no configuration was found after a timeout
+  private val scheduledTimeout = giveUpAfterTimeout()
```

Comment:
This is probably out of scope but it seems that the current code is a bit too complex for all that it's done.

For the timeout, can't we use a simpler `Future.firstCompletedOf` where we have the initial config set using an akka `pattern.after` to complete the future after the delay? And create a source using `Source.future` from the previous result, concating the streaming update, then using a `via` with the killswitch. This would eliminate the need for the timeout and the state kept in the killSwitch and the readyPromise.

Reply:
I think I see what you mean. Definitely worth a shot at cleaning up.

Reply:
I gave this a shot but everything I do makes it messier. Please feel free to take a look yourself once this is merged. :grinning:

Reply:
Sounds good 😂